crypto/hpke/hpke_util.c

```diff
@@ -192,9 +214,17 @@ const OSSL_HPKE_KEM_INFO *ossl_HPKE_KEM_INFO_find_random(OSSL_LIB_CTX *ctx)
     unsigned char rval = 0;
     int sz = OSSL_NELEM(hpke_kem_tab);
 
-    if (RAND_bytes_ex(ctx, &rval, sizeof(rval), 0) <= 0)
+    rval = HPKE_KEM_TAB_UB;
+    if (sz >= rval) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
         return NULL;
-    return &hpke_kem_tab[rval % sz];
+    }
+    while (rval >= sz) {
+        if (RAND_bytes_ex(ctx, &rval, sizeof(rval), 0) <= 0)
+            return NULL;
+        rval = rval % HPKE_KEM_TAB_UB;
```

Comment:
It might be useful to write a function to generate an unbiassed random integer in a range rather than repeating this loop in several places.  It's would be worthwhile as an addition to our internal headers (`internal/safe_math.h` or `crypto/rand.h` perhaps) so it can be used more widely.

We've got a BN function to do this (`BN_rand_range`) which is probably overkill here.  There is a brilliant algorithm detailed in [Fast Random Integer Generation in an Interval](https://arxiv.org/abs/1805.10941) by Daniel Lemire which avoids the modulus operation mostly.  This has since been improved upon to avoid modulus completely I believe.

This would also avoid needing the round up to a power of two stuff.

Reply:
Also nice. Will investigate. OTOH, while nice, I'm not sure if applications using HPKE and GREASEing (e.g. ECH or MLS) will really benefit directly - chances are in the end it'll be a better option to mimic what some popular browser does (once we know that), but you're likely right there's a useful utility function to be made here, so worth a look.

Reply:
I remember the _improved_ version: [An optimal algorithm for bounded random integers](https://github.com/apple/swift/pull/39143).  Division free and optimal in the amount of entry consumed.

Reply:
I've tossed together an implementation of a variant of the latter algorithm in #22499.
I'll polish it off tomorrow I expect.

Reply:
Even nicer! I'm guessing #22499 is likely to be merged before this, so meanwhile will copy in a local version of ``ossl_rand_uniform_uint32()`` and use that. (If some other plan is better, happy to hear what.)

Reply:
You could rebase this over the top of mine.

I will warn that the code hasn't been executed yet and will likely crash :)

Reply:
looks like it works pretty ok, will post a bit of test code to the other branch